server/gotenberg/gotenberg.go

```diff
@@ -0,0 +1,197 @@
+	if err != nil {
+		return false, err
+	}
+	if _, err := c.Get(url); err != nil {
+		return false, nil
+	}
+	return true, nil
+}
+
+// Convert converts file with given name and extension to PDF.
+// caller is responsible to Close() PDF stream after done.
+func (g *Gotenberg) Convert(name, extension string, file io.Reader) (pdf io.ReadCloser, err error) {
+	// check to see if given file extension is supported.
+	if !isSupported(extension) {
+		return nil, fmt.Errorf("file extension `%s` is not supported by the PDF server", extension)
+	}
+	// create a pipe and:
+	// - give the pw to multipart writer so it can start writing multipart data back while reading
+	//   the contents of file(io.Reader).
+	// - give the pr to HTTP request that will be made to Gotenberg. this way, as Gotenberg server
+	//   do reads, multipart writer can continue to write in sync.
+	//
+	// with this approach we have data cycled as a stream which reduces memory usage.
+	pr, pw := io.Pipe()
+	writer := multipart.NewWriter(pw)
+	// closer finishes multipart writer's work and closes the pipe with a regular EOF or an error.
+	// in both cases HTTP request to Gotenberg will be completed resulting without an error (in case of EOF)
+	// or cancelling with an error (in case of non-EOF).
+	closer := func(err error) {
+		writer.Close()
+		pw.CloseWithError(err)
+	}
+	// create a 'multipart file' and copy whole content of file as Gotenberg server continues to read.
+	go func() {
+		part, err := writer.CreateFormFile(multipartField, fmt.Sprintf("%s.%s", name, extension))
+		if err != nil {
+			closer(err)
+			return
+		}
+		_, err = io.Copy(part, file)
+		closer(err)
+	}()
+	conf := g.cloneConfing()
+	url, err := buildGotenbergURL(conf.addr, convertEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	// make an HTTP request to Gotenberg to initialize process.
+	req, err := http.NewRequest("POST", url, pr)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Content-Type", writer.FormDataContentType())
+	c := &http.Client{Timeout: conf.convertTimeout}
+	res, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	// check if Gotenberg is cool with the file we sent to see if it's gonna response back with a PDF data.
+	if res.StatusCode != http.StatusOK &&
+		res.StatusCode != http.StatusCreated {
+		defer res.Body.Close()
+		// request is somehow not successful:
+		// file content can be invalid or some timeout might be hitting set by Gotenberg's end or here in the request.
+		data, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, fmt.Errorf("error from Gotenberg with '%d' code: %s", res.StatusCode, string(data))
+	}
+	// we have Gotenberg willing to stream PDF data, give it to the caller so it can start reading.
+	return res.Body, nil
+}
+
+// buildGotenbergURL generates a Gotenberg API URL from given addr for endpoint.
+// TODO(ilgooz): maybe cache generated URLs in memory to improve performance.
```

Comment:
Can you elaborate on the potential performance concern here?

Reply:
Nothing much really. It would just save us from some additional cpu cycles. Nothing mandatory, I'll remove this TODO.

Reply:
done by done by 3a0f014